test/integration/query/selects.js

```diff
@@ -150,6 +150,78 @@ module.exports = function (knex) {
         });
     });
 
+    it('#4199 - adheres to optimizer hints', async function () {
+      const expectedErrors = {
+        mysql: {
+          code: 'ER_QUERY_TIMEOUT',
+          errno: 3024,
+          sqlMessage: 'Query execution was interrupted, maximum statement execution time exceeded',
+        },
+        mysql2: {
+          errno: 3024,
+          sqlMessage: 'Query execution was interrupted, maximum statement execution time exceeded',
+        },
+      }
+      if (!expectedErrors[knex.client.driverName]) {
+        return this.skip();
+      }
+      const baseQuery = knex('accounts')
+        .select('id', knex.raw('sleep(0.1)'))
+        .limit(2)
+      await expect(
+        baseQuery.clone()
+      ).to.eventually.be.fulfilled.and.to.have.lengthOf(2)
+      await expect(
+        baseQuery.clone().optimizerHint('max_execution_time(10)')
+      ).to.eventually.be.rejected.and.to.deep.include(expectedErrors[knex.client.driverName])
+    });
+
+    it('#4199 - ignores invalid optimizer hints', async function () {
```

Comment:
Neither Postres, Sqlite3 nor MSSQL support this kind of hints. So I at least test they don't interfere if used.

Reply:
MSSQL does, I think: https://docs.microsoft.com/en-us/sql/t-sql/queries/hints-transact-sql-query?view=sql-server-ver15

Everything that does not should be throwing an error.

Reply:
@kibertoad The problem with MSSQL is that it uses OPTION clause instead of `/*+ ... */` comment syntax, which works differently.

For instance we cannot just simply validate the hint for non-existence of `*/` within the hint to prevent a premature end of the hint as mentioned in https://github.com/knex/knex/pull/2815#discussion_r261689009. We would need to parse the hint itself to know whether there are not more `)` than `(` in it. Otherwise any code could leak into the query passing somehting like `.optimizerHint('HASH GROUP) injected sql (...')`.

Also as the OPTION clause is a normal SQL clause, it supports parameter bindings, while the comment hints syntax does not. So I'm not sure if it is good to mingle two so different syntexes into a single method. Possibly we could implement the OPTION clause using a separate method.

OTOH, we may let MSSQL use the OPTION clause for the hints, but without checking for injections within hints. And with forbidden parameter bindings for now, as adding parameter bindings will mess with the comment syntax. Assuming there are no proxy interpreters / middlewares / routers for MSSQL using the comments hint syntax. Otherwise we would prevent users from passing the comment hints to them when using MSSQL dialect.

Perhaps we could rename the `.optimizerHint(hint)` method to something like `.commentHint(hint)` to emphasize it is for hints passed via the `/*+` comment syntax regardless of used dialect.

Reply:
@martinmacko47 Then it sounds like current validation logic should be moved from generic querybuilder into MySQL querybuilder. Then we could implement it differently for MSSQL. Would that allow to use same `optimizerHint` for both?

Reply:
@kibertoad I guess we need to decide whether we want:

a) A semantic method for optimizer hints as they are supported in each dialect. That would be comment-like syntax for MySQL and Oracle, OPTION clause for MSSQL, and unsupported exception for Postgres and SQLite. With no way to pass comment-like hints for proxies / routers / whatever middlewares people may use.

b) A common method for any hints with comment-like syntax regardless they are consumed by the DB itself or any party on the way to DB.

I don't really have an opinion which one is more useful.

Reply:
It's a tough one to pick. On one heand one of the main goals of knex is to abstract away dialect differences between different DBs, so that swapping them works as transparently as possible. Now in this case one would obviously need to rewrite hints across dbs either way, but using unified syntax for that is still likely to be more familiar for people doing the change.

However, I see how that would be limiting the use of the hints for non-optimizer cases, hence I probably would vote for the second one, with very explicit documentation of what use-case it could be used.

Reply:
@kibertoad Then let's keep the comment-like syntax hints for all dialects and we will try to document it properly. Just perhaps let's name the method not `optimizerHint` so users won't confuse it with other types of optimizer hints in other dialects, but somehow else.

Reply:
MySQL documentation calls them "optimizer hint comments", so either `optimizerHintComment` or `hintComment` works for me.

Reply:
@kibertoad Renamed to `hintComment` in ad8b9a6. The other is too long IMHO.